src/server/main_service.cc

```diff
@@ -1126,6 +1129,52 @@ bool IsWatchingOtherDbs(DbIndex db_indx, const ConnectionState::ExecInfo& exec_i
   return false;
 }
 
+template <typename F> void IterateAllKeys(ConnectionState::ExecInfo* exec_info, F&& f) {
+  for (auto& [dbid, key] : exec_info->watched_keys)
+    f(MutableSlice{key.data(), key.size()});
+
+  for (const auto& scmd : exec_info->body) {
+    auto args = scmd.ArgList();
+
+    auto key_res = DetermineKeys(scmd.descr, args);
+    if (!key_res.ok())
+      continue;
+
+    auto key_index = key_res.value();
+
+    for (unsigned i = key_index.start; i < key_index.end; i += key_index.step)
+      f(args[i]);
+
+    if (key_index.bonus)
+      f(args[key_index.bonus]);
+  }
+}
+
+CmdArgVec CollectAllKeys(ConnectionState::ExecInfo* exec_info) {
+  CmdArgVec out;
+  out.reserve(exec_info->watched_keys.size() + exec_info->body.size());
+
+  IterateAllKeys(exec_info, [&out](MutableSlice key) { out.push_back(key); });
+
+  return out;
+}
+
+vector<bool> DetermineKeyShards(ConnectionState::ExecInfo* exec_info) {
+  vector<bool> out(shard_set->size());
```

Comment:
We should make this a bitset in the future. I assume we won't have more than ~256 shards